### PR TITLE
Postgres support array now & added a upsert function in tables.py

### PR DESCRIPTION
**Pull request by adhxtry**
**UPSERT**
```python
      Usage:
            db.tables['stocks_new_tb2'].upsert(
                date='2006-01-05',
                trans={
                    'type': 'BUY', 
                    'conditions': {'limit': '36.00', 'time': 'EndOfTradingDay'}, #JSON
                'tradeTimes':['16:30:00.00','16:30:01.00']}, # JSON
                symbol='RHAT', 
                qty=100.0,
                price=35.14,
                target='symbol')
            => Same Usage as `insert()` but with optional target argument.
```
**ARRAY IN POSTGRES**
Pass a list to kwarg and it converts to array, also when u select an postgres array it returns list assign to column named key, like:
[{'fruits': ['apple', 'grape']}]
insert, upsert and update all support postgres arrays.

### CODE REVIEW

File: aiopyql/table.py [hunk 1]
```diff
@@ -428,8 +428,13 @@ async def insert(self, **kw):
                 cols = f'{cols}, '
                 vals = f'{vals}, '
             cols = f'{cols}{col_name}'
+            #postgres array
+            if str(self.database) == 'postgres' and col.type == str and (kw[col_name][0], kw[col_name][-1]) == ('[',']'):
+                print('hi')
```
Comment (codemation) [inline aiopyql/table.py:433]: This is a friendly, but probably un-required print statement

File: aiopyql/table.py [hunk 2]
```diff
@@ -428,8 +428,13 @@ async def insert(self, **kw):
                 cols = f'{cols}, '
                 vals = f'{vals}, '
             cols = f'{cols}{col_name}'
+            #postgres array
+            if str(self.database) == 'postgres' and col.type == str and (kw[col_name][0], kw[col_name][-1]) == ('[',']'):
```
Comment (codemation) [inline aiopyql/table.py]:
```python
class Database 
..
....
    def __str__(self):
        return self.db_name

>> str(self.database) 
'db_name'

>> self.database.type in ['sqlite', 'mysql', 'postgres']
True
```

Comment (codemation) [reply aiopyql/table.py]: @AdhTri-001

File: aiopyql/table.py (hunk 1 above)
Comment (codemation) [reply aiopyql/table.py:433]: @AdhTri-001

File: aiopyql/table.py [hunk 3]
```diff
@@ -535,8 +587,12 @@ async def update(self, where: dict, **kw):
                 continue
             if len(cols_to_set) > 1:
                 cols_to_set = f'{cols_to_set}, '
+            #postgres array
+            if str(self.database) == 'postgres' and self.columns[col_name].type == str and (col_val[0], col_val[-1]) == ('[',']'):
```
Comment (codemation) [inline aiopyql/table.py]: @AdhTri-001  same as above

**Review by codemation — COMMENTED** [review]
@AdhTri-001 - Nice job so far, but few improvments that are needed before merging. 
- [ ] Fix self.database check type checks to correctly use self.database.type when checking for postgres arrays 
- [ ] Remove un-need print statments
- [ ] Provide Usage Examples so these can be attached to unittests

File: aiopyql/table.py (hunk 1 above)
Comment (adhxtry) [reply aiopyql/table.py:433]: Lmao that was for testing

**Comment by adhxtry** [comment]
Thanks, These are some tests i did with postgres:
image
this one with no primary key provided to `uspert()` :
image
This one with primary key provided:
image

And now tests for other variable types:
image
Returns in str form because of:
image
its text[]

great woho thats all for arrays.
now some `upsert()` test.
well all the above snaps proofs that upsert works perfectly fine, and the condition is that `PRIMARY KEY` must be provided.
thanks @codemation  you did awesome work.

File: aiopyql/table.py [hunk 4]
```diff
@@ -561,6 +616,7 @@ async def update(self, where: dict, **kw):
             return result
         except Exception as e:
             self.log.exception(f"Exception updating row for {self.name}")
+            print("\nQuery produced: ", query)
```
Comment (codemation) [inline aiopyql/table.py:614]: @AdhTri-001 
if printing is required, or deemed helpful, then it is better to use self.log.debug() or self.log.warning()

**Comment by codemation** [comment]
@AdhTri-001

The main unit test file is located [here](https://github.com/codemation/aiopyql/blob/master/tests/main_test.py). See if you can add a few upsert examples on the 'stocks', or 'employees' table, or 'keystore' ( probably the easiest). 

To test the tests afterwards, download the test_postgres.py and place in the same folder as main_test.py:

```python
python -m unittest test_postgres.py 
```

File: aiopyql/table.py (hunk 4 above)
Comment (adhxtry) [reply aiopyql/table.py:614]: Ok

**Pull request closed without merging**